convert2rhel/checks.py

```diff
@@ -16,13 +16,207 @@
+
+def _convert_to_int_or_zero(s):
+    try:
+        return int(s)
+    except ValueError:
+        return 0
+
+
+def _get_kmod_comparison_key(path):
+    """Create a comparison key from the kernel module abs path.
+
+    Converts /lib/modules/5.8.0-7642-generic/kernel/lib/a.ko.xz ->
+    kernel/lib/a.ko.xz
+
+    Why:
+        The standard kernel modules are located under
+        /lib/modules/{some kernel release}/.
+        If we want to make sure that the kernel package is present
+        on RHEL, we need to compare the full path, but because kernel release
+        might be different, we compare the relative paths after kernel release.
+    """
+    return "/".join(path.split("/")[4:])
+
+
+def get_installed_kmods():
+    """Get a set of kernel modules.
+
+    Each module we cut part of the path until the kernel release
+    (i.e. /lib/modules/5.8.0-7642-generic/kernel/lib/a.ko.xz ->
+    kernel/lib/a.ko.xz) in order to be able to compare with RHEL
+    kernel modules in case of different kernel release
+    """
+    try:
+        kernel_version, exit_code = run_subprocess("uname -r")
+        assert exit_code == 0
+        kmod_str, exit_code = run_subprocess(
+            'find /lib/modules/{kver} -name "*.ko*"'.format(
+                kver=kernel_version.rstrip("\n")
+            )
+        )
+        assert exit_code == 0
+        assert kmod_str
+    except (subprocess.CalledProcessError, AssertionError):
+        logger.critical("Can't get list of kernel modules.")
+    else:
+        return set(
+            _get_kmod_comparison_key(path)
+            for path in kmod_str.rstrip("\n").split()
+        )
+
+
+def _repos_version_key(pkg_name):
+    try:
+        rpm_version = KERNEL_REPO_RE.search(pkg_name).group("version")
+    except AttributeError:
+        logger.critical(
+            "Unexpected package:\n%s\n is a source of kernel modules.",
+            pkg_name,
+        )
+    else:
+        return tuple(
+            map(
+                _convert_to_int_or_zero,
+                KERNEL_REPO_VER_SPLIT_RE.split(rpm_version),
+            )
+        )
+
+
+def get_most_recent_unique_kernel_pkgs(pkgs):
+    """Return the most recent versions of all kernel packages.
+
+    When we scan kernel modules provided by kernel packages
+    it is expensive to check each kernel pkg. Since each new
+    kernel pkg do not deprecate kernel modules we only select
+    the most recent ones.
+
+    :type pkgs: Iterable[str]
+    """
+
+    pkgs_groups = itertools.groupby(
+        pkgs, lambda pkg_name: pkg_name.split(":")[0]
+    )
+    return (
+        max(distinct_kernel_pkgs[1], key=_repos_version_key)
+        for distinct_kernel_pkgs in pkgs_groups
+        if distinct_kernel_pkgs[0].startswith(("kernel", "kmod"))
+    )
+
+
+def get_rhel_supported_kmods():
+    """Return set of target RHEL supported kernel modules."""
+    repoquery_repoids_args = " ".join(
+        "--repoid " + repoid for repoid in system_info.default_rhsm_repoids
```

Comment:
This would work only in case of using RHSM, not the custom repos.

```suggestion
        "--repoid " + repoid for repoid in get_enabled_rhel_repos()
```

Create the following new function in systeminfo.py:
```python
get_enabled_rhel_repos():
    """Return a list of repoids containing RHEL packages.

    These are either the repos enabled through RHSM or the custom repositories passed though CLI.
    """
    return system_info.submgr_enabled_repos if not tool_opts.disable_submgr else tool_opts.enablerepo
```
Use this new function also on two other places:
- https://github.com/oamg/convert2rhel/blob/v0.18/convert2rhel/pkghandler.py#L122
- https://github.com/oamg/convert2rhel/blob/v0.18/convert2rhel/pkghandler.py#L637